flake-module.nix

```diff
@@ -95,16 +96,28 @@ in
                 description = ''Package overrides given new source path'';
                 default = { };
               };
-              overrides = mkOption {
-                type = functionTo (functionTo (types.lazyAttrsOf raw));
-                description = ''
-                  Overrides for the Cabal project
+              overrides =
+                let
+                  haskellOverlay = types.mkOptionType {
+                    name = "haskellOverlay";
+                    description = "Haskell overlay function";
+                    descriptionClass = "noun";
+                    check = lib.isFunction;
+                    merge = loc: defs:
+                      # TODO: What to do with loc?
```

Comment:
`loc` is for reporting the definition locations when doing type checking.
I don't think there's a function we could reuse for this, but Nixpkgs `lib/types.nix` has "example" code to work from.
The reason we can't reuse anything is because the interaction between type checking and laziness is quite subtle. The module system hasn't elaborated all the abstractions that might make this stuff obvious, and I don't know if that can be done at all and without making it slow and confusing.

In this case, I don't think we should use `composeManyExtensions`, because we can't really check the function bodies before know what the arguments are.
Another possible issue is that the ordering of the overlays matters, but is quite arbitrary in the module system. Ideally, modules are structured in such a way that the order in which the modules are processed does not matter, but for list and some overlays it does. We might want to detect that; see https://github.com/NixOS/nixpkgs/issues/215486

So this type is a bit of a rabbit hole. We'll want to go for incremental improvements. No checking at all is a valid first solution.

Reply:
> In this case, I don't think we should use `composeManyExtensions`, because we can't really check the function bodies before know what the arguments are.

But this is not important, right? The error message may not be helpful, but an error is thrown nonetheless, forcing the user to specify the overlay correctly.

> Another possible issue is that the ordering of the overlays matters

Is this blocker for implementing composable overlays that can be `import`'ed, then? Or, is the order not deterministic for simpler cases. Like take this one,

```nix
{ 
  haskellProjects.foo = {
    imports = [ inputs.myCompanyCommon.haskellFlakeProjectDeps; ];
    overrides = self: super:  { /* Local project overrides go here */ };
  };
};
```

Here, I would imagine the global dependency overries from `myCompanyCommon` repo to apply _first_ before the project `overrides` are applied, right? 

If the order is deterministic, I was thinking we could note the limitation of arbitrary order in the option docs.

Reply:
The ordering of the definitions passed to the merge function is deterministic in the sense that it only depends on the modules that are loaded and the implementation of the modules system, but I can't tell you what the order would be. It depend on details about how the import closure is computed and when and how often the module list is reversed. Also I don't want to know because it's a better idea for scalability to avoid depending on this implementation detail. If you move an override to another module, and the code breaks, wouldn't that be really really bad? It'd be unreliable and it stops you from refactoring the code, so that will become a mess.

I think we can just document it for now.

Reply:
By the way, even without this PR this problem remains - attrset merging (which functionTo relies on) too uses arbitrary order, but there it is not even using the overlay composition.